docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,298 @@
+ip netns add ns1
+
+# create the macvlan link attaching it to the parent host $DEV
+ip link add macvlan1 link $DEV type macvlan mode bridge
+
+# move the new interface macvlan1 to the new namespace
+ip link set macvlan1 netns ns1
+
+# bring the interface up
+# ns1 will later be used by libvirt to attach macvtaps to
+ip netns exec ns1 ip link set dev macvlan1 up
+```
+
+After that invocation, we have a host macvlan, with an IP address assigned. The
+CNI dhcp plugin comes with a minimalistic dhcp client, which we can use to do
+the DHCP renewals. An important side note here is, that the minimalistic daemon
+returns the acquired IP alog side with routing rules suggested by the DHCP
+server, but they are not applied, so the host is not modified.
+
+#### Prepare the VM device
+
+When a VM contains a network configuration like this:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: HostNetwork
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+```
+For every `HostNetwork`, `virt-handler` let's libvirt create a new macvtap
+attached to the one prepared macvlan in the initial setup step in the libvirt
+namespace.  Ee will use the CNI `dhcp` IPAM plugin with a provided or generated
+MAC address. Based on this MAC address and the IPAM plugin invocation we end up
+with the correct IP and MAC pair. The MAC address will then be used by libvirt
+as the mactap address. The IP can be used to add a route, to allow routing
+between the new guest and the host.
+
+Taking the macvlan name and the MAC address, we can now transform the
+HostNetwork entry into a libvirt XML fragment:
+
+```xml
+<interface type='direct'>
+  <mac address="fe:54:00:f7:78:6d"/>
+  <source dev='macvlan1' mode='bridge'/>
+</interface>
+```
+
+When the guest now tries to acquire an IP address with this mac address, it
+will get the same IP address. Based on that assumption we can create the route
+to allow the guest to talk to the node.
+
+```bash
+# Add an extra route which allows accessing the $DEV_IP through the master macvlan.
+# We need one of these routes per device.
+ip route add $DEV_SUBNET.11/32 via $DEV_SUBNET.13 dev macvlan0
+
+# ping from container to host
+# done by the ipam dhcp client
+ip netns exec ns1 ping $DEV_SUBNET.13
+```
+
+In this scenario, we can keep the IP of our main interface in place, and only
+have to add one additional route per acquired IP for a VM. All existing routes
+can stay untouched. In theory, moving the IP of the main interface to `macvlan0`
+would also be possible, but that would require extensive route rewrites which
+will likely cause conflicts with other components involved.
+
+### Bridge Scenario (tap)
+
+Normally with the Kubernetes overlay networks, bridges are not required or
+configured. We would have to configure the bridge based on an existing device.
+That is very hard to achive, since we don't own the host network.
```

Comment:
Is it such a hard requirement to place: "for using nested node, make sure your master host device is a bridge". Creating `virbr` and attaching it to `eth0` can happen during host deployment (day 1) or via a specialized pod running in the host network namespace.

Reply:
We're trying our best to avoid additional host setup for KubeVirt.  The goal to to allow someone with an existing k8s cluster to be able to drop in KubeVirt with no additional work.  That's why we're willing to absorb some additional complexity to avoid using a bridge.